File: scratch_smart_diff.py
import json
import urllib.request
import re

def clean(text):
    return re.sub(r'[^\u0621-\u064A]', '', text).replace('ي', 'ی').replace('ے', 'ی').replace('ا', 'ا').replace('أ', 'ا').replace('إ', 'ا').replace('آ', 'ا').replace('ة', 'ه')

# Load Qalon
with open('assets/data/output.json', 'r', encoding='utf-8') as f:
    data = json.load(f)

qalon_ayahs = []
for page in data:
    # Handle list of pages or list of lists
    pages = page if isinstance(page, list) else [page]
    for p in pages:
        for a in p.get('ayahs', []):
            if a['surah'] == 18:
                qalon_ayahs.append(clean(a['text']))

# Load Hafs
with urllib.request.urlopen('https://api.alquran.cloud/v1/surah/18') as url:
    hafs_data = json.loads(url.read().decode())

hafs_ayahs = [clean(a['text']) for a in hafs_data['data']['ayahs']]
# Strip basmalah
if hafs_ayahs[0].startswith('بسماللهالرحمنالرحیم'):
    hafs_ayahs[0] = hafs_ayahs[0][len('بسماللهالرحمنالرحیم'):]

q = 0
h = 0
merges = []
while q < len(qalon_ayahs) and h < len(hafs_ayahs):
    if qalon_ayahs[q] == hafs_ayahs[h]:
        q += 1; h += 1; continue
        
    # Check if Q combines H and H+1
    if h + 1 < len(hafs_ayahs):
        combined = hafs_ayahs[h] + hafs_ayahs[h+1]
        # Allow slight differences
        if abs(len(qalon_ayahs[q]) - len(combined)) < 10:
            merges.append(q + 1) # 1-based index
            q += 1; h += 2; continue
            
    # Try 3 ayahs
    if h + 2 < len(hafs_ayahs):
        combined = hafs_ayahs[h] + hafs_ayahs[h+1] + hafs_ayahs[h+2]
        if abs(len(qalon_ayahs[q]) - len(combined)) < 10:
            merges.append(q + 1)
            q += 1; h += 3; continue

    # Fallback to simple skip if nothing matches exactly (rare)
    q += 1; h += 1

print("Merges found at Qalon Ayah numbers:", merges)

File: scratch_smart_diff.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;

String clean(String text) {
  return text
      .replaceAll(RegExp(r'[^\u0621-\u064A]'), '')
      .replaceAll('ي', 'ی')
      .replaceAll('ے', 'ی')
      .replaceAll('أ', 'ا')
      .replaceAll('إ', 'ا')
      .replaceAll('آ', 'ا')
      .replaceAll('ة', 'ه');
}

void main() async {
  final file = File('assets/data/output.json');
  final data = json.decode(file.readAsStringSync()) as List<dynamic>;
  final flattened = <dynamic>[];
  for (final item in data) {
    if (item is List) { flattened.addAll(item); } else { flattened.add(item); }
  }
  
  final qalonAyahs = <int, String>{};
  for (final page in flattened) {
    for (final a in (page['ayahs'] as List)) {
      if (a['surah'] == 21) {
        qalonAyahs[a['ayah'] as int] = clean(a['text'] as String);
      }
    }
  }

  final response = await http.get(Uri.parse('https://api.alquran.cloud/v1/surah/21'));
  final hafsData = json.decode(response.body);
  final hafsAyahs = <int, String>{};
  for (final a in hafsData['data']['ayahs']) {
    hafsAyahs[a['numberInSurah'] as int] = clean(a['text'] as String);
  }

  if (hafsAyahs[1]!.startsWith('بسماللهالرحمنالرحیم')) {
    hafsAyahs[1] = hafsAyahs[1]!.substring('بسماللهالرحمنالرحیم'.length);
  }

  int q = 1;
  int h = 1;
  final merges = <int>[];
  
  while (q <= qalonAyahs.length && h <= hafsAyahs.length) {
    if (qalonAyahs[q] == hafsAyahs[h]) {
      q++; h++; continue;
    }
    
    if (h < hafsAyahs.length) {
      String combined = hafsAyahs[h]! + hafsAyahs[h+1]!;
      if ((qalonAyahs[q]!.length - combined.length).abs() < 10) {
        merges.add(q);
        q++; h += 2; continue;
      }
    }
    
    if (h < hafsAyahs.length - 1) {
      String combined = hafsAyahs[h]! + hafsAyahs[h+1]! + hafsAyahs[h+2]!;
      if ((qalonAyahs[q]!.length - combined.length).abs() < 10) {
        merges.add(q);
        q++; h += 3; continue;
      }
    }
    
    // Check if Hafs combines Q and Q+1 (split in Qalon)
    if (q < qalonAyahs.length) {
      String combinedQ = qalonAyahs[q]! + qalonAyahs[q+1]!;
      if ((combinedQ.length - hafsAyahs[h]!.length).abs() < 10) {
        print('SPLIT FOUND: Hafs $h is split into Qalon $q and ${q+1}');
        q += 2; h++; continue;
      }
    }
    
    print('Failed to match at Q=$q H=$h');
    q++; h++;
  }
  
  print('Merges found at Qalon Ayah numbers: \$merges');
}
